### ISSUE TITLE
[FEATURE REQ] Cache count configuration in Eventhubs SDK

### ISSUE THREAD

**Issue by chuckz1321** [Azure/azure-sdk-for-java#30996]
**Is your feature request related to a problem? Please describe.**
As tested, Java Eventhubs SDK may cache specific count of events and will relased the memory when the count reach a threshold. Correct me if I misunderstand. Because I find a parameter in [C# SDK](https://docs.microsoft.com/en-us/dotnet/api/azure.messaging.eventhubs.consumer.readeventoptions.cacheeventcount?view=azure-dotnet). 

Also, here's my test. The main function just have a processor and it will just print some info to console. I use Jconsole to monitor the memory. 
When message body is "hello", and I set a small prefetch count. 
![image](https://user-images.githubusercontent.com/33536264/190605475-e707efb2-0338-4b4b-aa31-ba1baf5e2c56.png)
![image](https://user-images.githubusercontent.com/33536264/190605555-2193271f-0432-452f-be03-b6324b82f21c.png)

When the message body is larger, the memory used increases. 
![image](https://user-images.githubusercontent.com/33536264/190605581-750ef943-57ce-4453-a01f-89a73086a769.png)
Customer may start multiple threads to consume messages in one Java process. And I assume if the message body is large like 0.5MB or larger, it will cause OOM based on the limited heap size configuration.
 
**Describe the solution you'd like**
So could we please add the controle to the cache size? If the memory raise is not caused by my consumption, please correct it and glad to hear your insights.

Feel free to ping me on Teams. chenzhang@microsoft.com

**Describe alternatives you've considered**

**Additional context**

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [ ] Description Added
- [ ] Expected solution specified

**Comment by chuckz1321** [comment]
Hi team, any insights on this issue?

**Comment by alzimmermsft** [comment]
@conniey, @anuchandy are there any updates for this?

**Comment by conniey** [comment]
Hey,

> As tested, Java Eventhubs SDK may cache specific count of events and will released the memory when the count reach a threshold. Correct me if I misunderstand. 

Based on your configuration, the Event Hub client library will cache the _prefetch_ count events, so 500 events.  It fetches more events when a portion of the events have been consumed by the caller and fills the prefetch queue back to the prefetch count.

We stop referencing the underlying AMQP message when it is successfully decoded.  We don't control objects that are garbage collected.

There are many factors that can increase memory usage.  The prefetch number and message size can affect it.

Before limiting your heap size, I would run your application for some time with a message of average size and see the max heap size trend when it stabilises.

**Comment by chuckz1321** [comment]
I am confused about the cache count and the value of prefetch parameter. You can compare the first and second pic. They received same messages(I mean the event body). The only difference is the prefetch count, first one is 500 and 20 for second one. And their peek memory values are all 100MB+

I noticed you said the reference of the event will be released when the processor complete the messages. But from this point, I highly suspect some variables still refer to it. :(

**Comment by chuckz1321** [comment]
Do you need the code? I can share it.

**Comment by chuckz1321** [comment]
@conniey Any update?

**Comment by anuchandy** [comment]
The prefetch configures the events to be fetched from the broker and buffered. The batchSize controls draining of this buffer. As the draining happens, the library communicates with the broker and tries to keep the buffer full.

Btw, when measuring, you want to ensure the same setup - both test cases got event hubs with enough events accumulated to keep providing events to SDK as soon as asked so library won't be idle, start measuring after JVM warmup (usually ~2-3 min of warmup up), same non-conjected network e.g. an Azure VM

When the event size increases, it is expected to see a positive delta in memory usage. Multiple factors are involved, e.g., The Heap for the buffer increases because of the increase in event size stored in each buffer slot. The operation of the reading and decoding event body must allocate a byte [] array depending on the event size. Other factors internal allocation in JVM IO APIs (read events bytes from n/w buffer to user space), the allocation from low-level AMQP library, etc.

**Comment by anuchandy** [comment]
The prefetch configures the events to be fetched from the broker and buffered. The batchSize controls draining of this buffer. As the draining happens, the library communicates with the broker and tries to keep the buffer full.

When the event size increases, it is expected to see a positive delta in memory usage. Multiple factors are involved, e.g., The Heap for the buffer increases because of the increase in event size stored in each buffer slot. The operation of the reading and decoding event body must allocate a byte [] array depending on the event size. (Other factors internal allocation in JVM IO APIs, the allocation from low-level AMQP library, etc.)

-----

If the goal is to identify how much memory to reserve for the application - measure and limit Heap accordingly. To expand on what [@conniey](https://github.com/conniey) shared earlier, a common way to measure is -

Run the application until it has reached a steady state. E.g., the application and JVM have loaded anything it caches [e.g., domain objects, class types, static instances, object pools], has created a maximum number of connections, and has a peak expected load). Now force a full GC using tools such as JVisualVM / JConsole. Check how much memory is occupied after the full GC. You want to size the Heap such that only 30% is occupied after the full GC; use this value to set the max heap size (-Xmx). Size each container memory to have an "additional ~1 GB" memory for the "non-heap" need for the JVM instance.
 
In a nutshell, measure the application for a long duration with messages of expected average size and see the max heap size trend when it stabilizes, then only limit Java Heap based on the observed value. You want to run your application against an Event Hub with enough messages to keep the SDK busy during the measuring.
